### PR TITLE
feat(all-events) add project type to column mapping

### PR DESCRIPTION
**Pull request by DominikB2014**
fixes PERF-1803

Adds a mapping between the platformCategory and columns, to display more relevant data. The full column mapping can be found in the spreadsheet on the ticket.

Note: InstallerStore, isSideLoaded and attachments is not included as it is not a field in discover, it will be looked into in the future

### CODE REVIEW

File: static/app/utils/platform.tsx
Comment (DominikB2014) [inline static/app/utils/platform.tsx:26]: If there is a cleaner way to do this, please let me know. I did it like this as it is consistent to the below. I'm also proposing removing `isMobilePlatform` as this function would provide the same functionality.

Comment (mjq-sentry) [reply static/app/utils/platform.tsx:26]: Since this function is essentially converting the string name of an enum value to that value, I'm sure there's an easier way without exhaustively testing every constant. I haven't done enough TypeScript to be sure but take a look at e.g. https://stackoverflow.com/a/42623905 which hopefully has some leads.

File: static/app/utils/platform.tsx
Comment (udameli) [inline static/app/utils/platform.tsx:15]: you can just create a map of string to enum value and then look up enums instead

Comment (udameli) [reply static/app/utils/platform.tsx:15]: something like `{ backend: PlatformCategory.backend }`

File: static/app/utils/platform.tsx
Comment (udameli) [reply static/app/utils/platform.tsx:26]: It's not blocking functionality-wise but it's an easy clean up and it'd be good to address it in this PR for overall code quality

Comment (DominikB2014) [reply static/app/utils/platform.tsx:26]: I agree! If there was no easy solution then I would be okay to address in the future.

File: static/app/utils/platform.tsx
Comment (DominikB2014) [reply static/app/utils/platform.tsx:15]: Great suggestion, i'll give that a go!

Comment (DominikB2014) [reply static/app/utils/platform.tsx:15]: Hmm, this function is meant to map a platform like `javascript`, `javascript-react`, `node` etc, to its corresponding category, such as `backend`, `frontend`. We already have a readonly array of all the platforms in a category, i.e `frontend = ['javascript', 'javascript-react' ...]`. Maybe something like this is cleaner
```
 const platformMap = {};
  mobile.forEach(mobilePlatform => platformMap[mobilePlatform] = PlatformCategory.mobile)
  frontend.forEach(frontendPlatform => platformMap[frontendPlatform] = PlatformCategory.frontend)
...
return platformMap[platform]
  ```

Comment (DominikB2014) [reply static/app/utils/platform.tsx:15]: And then we can store it outside the function so it doesn't get recomputed everytime

Comment (DominikB2014) [reply static/app/utils/platform.tsx]: I pushed a change, let me know what you think, Thanks for the suggestions!

File: static/app/utils/platform.tsx
Comment (0Calories) [inline static/app/utils/platform.tsx:29]: This works, but for the sake of readability, how about we just use the function to logically convert the platform to the enum instead of creating a map?

Something like this for example:

```
function platformToCategory( ... ) {
	if (mobile.includes(platform)) {
		return PlatformCategory.MOBILE
	}

	if (frontend.includes(platform)) {
		return PlatformCategory.FRONTEND
	}

	...etc
}
```

File: static/app/data/platformCategories.tsx
Comment (0Calories) [inline static/app/data/platformCategories.tsx:7]: enum values are typically denoted with all caps, i.e `FRONTEND`, `MOBILE`, we should change this here to be consistent with the rest of the codebase

File: static/app/utils/platform.tsx
Comment (DominikB2014) [reply static/app/utils/platform.tsx:29]: This is actually how it originally was and was because mobile/frontend arrays are these readonly constants requiring you to do some funky string array conversion like this https://github.com/getsentry/sentry/blob/35d73d9291e36806708e98387cf68c6c8af517af/static/app/utils/platform.tsx#L49

After previous code reviews regarding that, this is what I did as an alternative

File: static/app/data/platformCategories.tsx
Comment (DominikB2014) [reply static/app/data/platformCategories.tsx:7]: We should have a linting rule for this

Comment (0Calories) [reply static/app/data/platformCategories.tsx:7]: True, you should suggest it to the frontend channel! 😛

File: static/app/utils/platform.tsx
Comment (0Calories) [reply static/app/utils/platform.tsx:29]: lol honestly it seems like the root of the problem here is that these `mobile`, `frontend`, etc arrays are initialized as const string arrays when they should've just been enums to begin with, but it's going to be annoying to change that now. 

With how it's setup I personally don't find much issue with the original method via `return ([...mobile] as string[]).includes(platform);`. I'm not a huge fan of the map construction method since it feels a bit unnecessarily verbose but I'll leave the final decision up to you

Comment (DominikB2014) [reply static/app/utils/platform.tsx:29]: Tbh, I think the original method was a bit more readible, but I believe the map construction would be more performant (hash map vs array lookup), definitely not a concern for 100 items. I'll change it back!

Comment (0Calories) [reply static/app/utils/platform.tsx:29]: Yep, it's constant time in both cases because we have a set number of elements

Comment (DominikB2014) [reply static/app/utils/platform.tsx]: made the change!

File: static/app/views/organizationGroupDetails/allEventsTable.tsx
Comment (0Calories) [inline static/app/views/organizationGroupDetails/allEventsTable.tsx:147]: Instead of this, I think we should just create another entry for `PlatformCategory.SERVERLESS` within `categoryToColumnMap`. It will have the same columns but I think it makes more sense to keep the structure of passing the `platformCategory` to the map instead of having some intermediary logic that re-assigns the value

File: static/app/views/organizationGroupDetails/allEventsTable.tsx
Comment (0Calories) [inline static/app/views/organizationGroupDetails/allEventsTable.tsx:103]: I don't see `minidump` as an entry in the `fields` array, is there some hidden logic somewhere assigning this?

File: static/app/views/organizationGroupDetails/allEventsTable.tsx
Comment (0Calories) [inline static/app/views/organizationGroupDetails/allEventsTable.tsx:119]: A stricter way to type this would be like this:

```suggestion
  const categoryToColumnMap: Record<PlatformCategory, ColumnInfo> = {
```

One caveat is that you'll have to add an entry for `[PlatformCategory.SERVERLESS]` or the linter will complain, but that's proof that TS is properly doing its job here 😄

**Review by 0Calories — COMMENTED** [review]
LGTM but I got a few more comments/questions before I smash the approve button

File: static/app/views/organizationGroupDetails/allEventsTable.tsx
Comment (DominikB2014) [reply static/app/views/organizationGroupDetails/allEventsTable.tsx:103]: Yeah, minidump is a special field handled by the table itself

File: static/app/views/organizationGroupDetails/allEventsTable.tsx
Comment (DominikB2014) [reply static/app/views/organizationGroupDetails/allEventsTable.tsx:119]: I like this! forgot about using record, thanks!

File: static/app/views/organizationGroupDetails/allEventsTable.tsx
Comment (DominikB2014) [reply static/app/views/organizationGroupDetails/allEventsTable.tsx:147]: I do agree the reassignment is not ideal,

We want to ensure serverless and backend always get the same columns, even down the line as we modify columns, to prevent accidental changes I'd prefer not duplicating. How about adding this
`categoryToColumnMap[PlatformCategory.SERVERLESS] = categoryToColumnMap[PlatformCategory.BACKEND]`?

File: static/app/views/organizationGroupDetails/allEventsTable.tsx
Comment (0Calories) [reply static/app/views/organizationGroupDetails/allEventsTable.tsx:103]: I dug into the events table and I see that it's manually getting pushed in there, would we be able to push the title along with the field instead of assigning it here?

Comment (DominikB2014) [reply static/app/views/organizationGroupDetails/allEventsTable.tsx:103]: Not without more changes, as fields get added to the event view and become part of the discover query, fetching minidumps is not through discover.

Comment (DominikB2014) [reply static/app/views/organizationGroupDetails/allEventsTable.tsx:103]: Actually mb, I might have misunderstood what you meant. I'd want to push it here, in the `allEventsTable.tsx`

File: static/app/views/organizationGroupDetails/allEventsTable.tsx
Comment (DominikB2014) [reply static/app/views/organizationGroupDetails/allEventsTable.tsx]: Hmm, this actually doesn't work straight up like this, lemme dig into it some more
image

File: static/app/views/organizationGroupDetails/allEventsTable.tsx
Comment (DominikB2014) [reply static/app/views/organizationGroupDetails/allEventsTable.tsx:147]: Yeah, lets do that, thanks @0Calories !
That solves that typing issue I just mentioned, with mising '3'

File: static/app/views/organizationGroupDetails/allEventsTable.tsx
Comment (0Calories) [reply static/app/views/organizationGroupDetails/allEventsTable.tsx]: Yep it's because you have to add `[PlatformCategory.SERVERLESS]` as an entry, which is `3` in the enum.

Comment (DominikB2014) [reply static/app/views/organizationGroupDetails/allEventsTable.tsx]: Would be nice if vscode showed the enum instead of the value, got confused for a sec lol, but it makes sense.

Comment (0Calories) [reply static/app/views/organizationGroupDetails/allEventsTable.tsx]: Agreed, I only know this because I've been confused by this in the past haha

One way to make it cleaner when linting is use string values in the enum, like 
```
enum PlatformCategory = {
  MOBILE = 'mobile',
  FRONTEND = 'frontend',
  ...
}
```

**Review by 0Calories — APPROVED** [review]
noice 🎉

**Pull request merged**
